### ISSUE TITLE
Adding new collection creates Integrity Error, UNIQUE constraint failed

### ISSUE THREAD

**Issue by kristinriebe** [wagtail/wagtail#9208]
### Issue Summary
After renaming a collection via the admin interface, adding a new collection very often results in an IntegrityError, violating a UNIQUE constraint (or Duplicate key error).  
Possible reason: Since Wagtail 2.11, collections are created in alphabetical order (within each parent), which is ensured by a path reordering each time a new collection is created. (Which is a nice feature!)  
However, it is possible to rename existing collections (e.g. via the admin interface), but then the path won't be updated accordingly, thus breaking the alphabetical order. This wouldn't necessarily be a problem, but Django Treebeard seems to very much rely on this alphabetical ordering, so when the next collection is being created, it can happen that Treebeard tries to move collections to paths that already exist.

### Steps to Reproduce

I have reproduced the problem with a fresh Wagtail installation 2.15.5, using Python 3.6, Django 3.2, Wagtail 2.15.5, SQLite:

1. Start a fresh Wagtail installation by following the steps at https://docs.wagtail.org/en/v2.15.5/getting_started/tutorial.html, create a superuser and start the server.
2.  Login to the Wagtail admin interface and go to settings, collections. (http://127.0.0.1:8000/admin/collections)
3.  Create for example the following collections, as children of the root collection:
    - "Amphibians" 
    - "Fish"
    - "Crocodiles"
    The "Crocodiles" collection will be sorted correctly between "Amphibians" and "Fish". The current path values are:
    ```SQL
    sqlite> SELECT id, name, path FROM wagtailcore_collection ORDER BY name;
    2|Amphibians|00010001
    4|Crocodiles|00010002
    3|Fish      |00010003
    1|Root      |0001
    ```
4. Rename the "Crocodiles" collection to "Reptiles". The paths are not being updated to reflect this change.
5. Add new collection "Mammals" (as child of "Root" again). Now the IntegrityError happens.

### Error details
The shown error message is:
```
IntegrityError at /admin/collections/add/
UNIQUE constraint failed: wagtailcore_collection.path
```
I think this fails because Treebeard still assumes an alphabetical ordering (which is not true anymore) and thus tries to move a collection to an already existing path. The paths are now:
```SQL
sqlite> SELECT id, name, path FROM wagtailcore_collection ORDER BY name;
2|Amphibians|00010001
3|Fish      |00010003
4|Reptiles  |00010002
1|Root      |0001
```
And the offending SQL statement attempted by Treebeard for reordering the collections is:
```SQL
# env/lib/python3.6/site-packages/treebeard/mp_tree.py, line 145, in run_sql_stmts,
# cursor.execute(sql, vals)
# Class treebeard.mp_tree.MP_AddSiblingHandler

UPDATE "wagtailcore_collection" SET path='00010003'||SUBSTR(path, 9) WHERE path LIKE '00010002%';
```

Adding a collection named "Birds" would work, because then "Fish" would be moved first. 

### Work around / Suggestions

Using `python manage.py fixtree --full`fixes the collection paths and restores the alphabetical order.  
Basically, this needs to be done whenever a collection was renamed or every time before a collection is added. Could this be automated?  
If not, or as a quick fix, it would be helpful to mention in the documentation that the management command `fixtree`also fixes collection paths and that it should be applied after renaming collections (e.g. at https://docs.wagtail.org/en/v2.15.5/reference/management_commands.html#fixtree).  
It would also be awesome if Wagtail would not crash with a server error in production when this error happens, but instead could catch it beforehand and give a more meaningful erorr message. (Something like "Ask your web admin to fix the collection tree").

### Technical details

-   Python version: 3.6.7
-   Django version: 3.2.15
-   Wagtail version: 2.15.5

**Comment by richie-blake** [comment]
I've seen this behaviour on Wagtail 4.0.4 also. fixtree command resolved the issue and corrected the ordering.

**Comment by andre-fuchs** [comment]
I’ve had this problem on two instances of Wagtail, at least. Very confusing for end users as the only feedback is an "internal server error".

**Comment by tognee** [comment]
Could this be related to this issue?

https://github.com/django-treebeard/django-treebeard/issues/289

**Comment by gasman** [comment]
@tognee I don't think so - that issue is about performing multiple tree operations without reloading the in-memory node in between. According to @kristinriebe's analysis, the problem here originates from the path not being updated when the collection is renamed in the edit view - there's only one database operation at play there.

I think the root cause is summed up by https://django-treebeard.readthedocs.io/en/latest/mp_tree.html#treebeard.mp_tree.MP_Node.node_order_by:

> When enabled, all tree operations will assume this ordering.

A simple save to update the collection's name probably isn't considered a "tree operation", even though it can break the ordering that future tree operations rely on. To fix this, either Treebeard would have to intercept the `save` operation (which would probably be very disruptive to retro-fit into Treebeard's existing behaviour) or the edit view has to do the tree path housekeeping itself. This is effectively what @andre-fuchs's solution is doing, but doing a full `fixtree` operation is unnecessarily heavyweight - a single 'move' operation on the updated collection should be sufficient (as long as Treebeard's `move` implementation doesn't trip us up by assuming nodes are correctly ordered).
